pkg/client/unversioned/helper.go

```diff
@@ -177,65 +180,84 @@ func MatchesServerVersion(client *Client, c *Config) error {
 
 // NegotiateVersion queries the server's supported api versions to find
 // a version that both client and server support.
-// - If no version is provided, try registered client versions in order of
-//   preference.
-// - If version is provided, but not default config (explicitly requested via
-//   commandline flag), and is unsupported by the server, print a warning to
-//   stderr and try client's registered versions in order of preference.
-// - If version is config default, and the server does not support it,
-//   return an error.
-func NegotiateVersion(client *Client, c *Config, version string, clientRegisteredVersions []string) (string, error) {
-	var err error
-	if client == nil {
-		client, err = New(c)
-		if err != nil {
-			return "", err
-		}
-	}
-	clientVersions := sets.String{}
-	for _, v := range clientRegisteredVersions {
-		clientVersions.Insert(v)
+// - If no version is provided, try the version specified in the Config c. If that
+//   fails, try registered client versions in order of preference.
+// - If version is provided, but not the on in Config c, and is unsupported by
+//   the server, print a warning to stderr and try client's registered versions
+//   in order of preference.
+// - If version is provided and equals the one specified in the Config c, and
+//   the server does not support it, return an error.
+func NegotiateVersion(c *Config, group string, version string, clientRegisteredGroupVersions []string) (string, error) {
```

Comment:
I'd like to add in that the term `GroupVersion` really just makes this hard to understand. I keep wanting to think it's a group's version instead of some weird amalgamation of group name and possibly version.